### PR TITLE
MOVEONLY: Move common init code to init/common

### PR DESCRIPTION
**Pull request by ryanofsky**
This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/projects/10).

---

This change is move-only and can be easily reviewed with `--color-moved=dimmed_zebra`. The moves are needed to avoid duplicating common init code between different binaries (`bitcoin-node`, `bitcoin-wallet`, etc) in #10102. In #10102, each binary has it's own init file (`src/init/bitcoin-node.cpp`, `src/init/bitcoin-wallet.cpp`) so this PR moves the common code to `src/init/common.cpp`.

### CODE REVIEW

**Comment by ryanofsky** [comment]
Updated cade20c773e554f35ba148c4ee5243d1bda19805 -> 66bbbc09e9bb44546ac47829384cd3b088c5efd7 ([`pr/initc.1`](https://github.com/ryanofsky/bitcoin/commits/pr/initc.1) -> [`pr/initc.2`](https://github.com/ryanofsky/bitcoin/commits/pr/initc.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/initc.1..pr/initc.2)) to fix hidden_args CI errors https://cirrus-ci.com/task/4601873798791168?logs=ci#L3751 https://cirrus-ci.com/task/5446298728923136?logs=ci#L2663 https://cirrus-ci.com/task/5516667473100800?logs=ci#L4299 https://cirrus-ci.com/task/6079617426522112?logs=ci#L2817 https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/38776308

**Review by promag — COMMENTED** [review]
Code review ACK 66bbbc09e9bb44546ac47829384cd3b088c5efd7.

Verified commit by commit, with `--color-moved=dimmed-zebra`. Moved code is tidy up in nice functions. I don't observe changes in behavior.

**Comment by practicalswift** [comment]
Concept ACK

Nice with the `init` namespace too: it makes it easier to reason about encapsulation and more specifically it allows for finding all non-`init` call sites just by doing `git grep "init::"`.

More generally I think it would make sense to gradually introduce more of this type of namespace use in our project.

**Review by KenG-pro — APPROVED** [review]

File: src/init/common.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <key.h>
+#include <logging.h>
+#include <pubkey.h>
+#include <random.h>
```
Comment (kiminuo) [inline src/init/common.cpp:15]: Should `#include <crypto/sha256.h>` be here too or not?

File: src/init/common.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <memory>
+
+static std::unique_ptr<ECCVerifyHandle> globalVerifyHandle;
+
+namespace init {
+void SetGlobals()
+{
+    // Initialize elliptic curve code
+    std::string sha256_algo = SHA256AutoDetect();
```
Comment (MarcoFalke) [inline src/init/common.cpp:28]: unrelated 000e94b26c7340dfc834299d75848c34c73b3128: comment seems wrong? Maybe just remove it?

File: src/init.cpp [hunk 3]
```diff
@@ -1149,7 +1125,7 @@ bool AppInitSanityChecks()
     init::SetGlobals();
 
     // Sanity check
-    if (!InitSanityCheck())
+    if (!init::SanityChecks())
```
Comment (MarcoFalke) [inline src/init.cpp:1078]: e01dfceb3c2f355e8b9979eb6aa6c41550c37787: Could add `{` while touching this line?

File: src/init.cpp [hunk 4]
```diff
@@ -1149,7 +1125,7 @@ bool AppInitSanityChecks()
     init::SetGlobals();
 
     // Sanity check
```
Comment (MarcoFalke) [inline src/init.cpp:1077]: e01dfceb3c2f355e8b9979eb6aa6c41550c37787: Could remove comment that just repeats the function name that is called?

File: src/init/common.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+class ArgsManager;
+
+namespace init {
+void SetGlobals();
+void UnsetGlobals();
+/** Sanity checks
+ *  Ensure that Bitcoin is running in a usable environment with all
+ *  necessary library support.
+ */
+bool SanityChecks();
```
Comment (MarcoFalke) [inline src/init/common.h:20]: e01dfceb3c2f355e8b9979eb6aa6c41550c37787: Could remove comment that just repeats the function name?

```suggestion
/**
 *  Ensure a usable environment with all
 *  necessary library support.
 */
bool SanityChecks();
```

File: src/init.cpp [hunk 6]
```diff
@@ -22,6 +22,7 @@
 #include <httpserver.h>
 #include <index/blockfilterindex.h>
 #include <index/txindex.h>
+#include <init/common.h>
 #include <interfaces/chain.h>
 #include <interfaces/node.h>
 #include <key.h>
```
Comment (MarcoFalke) [inline src/init.cpp:28]: 000e94b26c7340dfc834299d75848c34c73b3128: Could remove pubkey/key include, now that they are no longer used?

**Review by maflcko — APPROVED** [review]
Left some unrelated doc feedback to remove nonsense comments like

```cpp
// Sanity checks
SanityChecks();
```

--color-moved=dimmed-zebra ACK  66bbbc09e9bb44546ac47829384cd3b088c5efd7 😹

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

- --color-moved=dimmed-zebra ACK  66bbbc09e9bb44546ac47829384cd3b088c5efd7 😹
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjyPQwAltkH6kZvb6r7SNqiODJqjD0Ta96ZqC0rKjnSwMrLmObSuMu8OgWJ1zJW
c2q68lGd1PGZHvDiIrlJu6939yGby5c+hj84iCkgJoL720ZrkXVQynPjtkzcOcEc
vfa5VGH18547lkFyih1LYepSJ9MBnU+XH4WYZJMyeu75NVEc3OZTCj3xG1U519n4
NyAdTJ8WK9KLZF8rYRP0/sBgXBawZtjaS43YSrdz061Ht8kTlU/YqQinbRs4mm0K
EdRybSUXzDeFxGHCq2Kz6vuooSDasiTZUoVf66crNDN5XEIpmI2NgBDt8SjssQIg
RQQ9vNGoHDqM4b0lI+zWc9kkply/82ZL1TJP1HBYy/gdVh+Z5M2L3q9RVOt315EJ
VZnfl2fp2NNJoqcP/mLx2ZF4eZ5hNWJLiFw5bdDvmnbb9Gz3rS4lWg+0MrW3P6Bs
XSb+eDLPgbluTlH5Azkn/j2KtuZf5ODk57igT6hg4ZxVwrspkoXIFckgitA3ddXu
eN/gtak+
=4TTx
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `27c6f524949cc5fa807c0a855609234ba5afe961ff88e1b28074374eea19342f  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010827c6f524949cc5fa807c0a855609234ba5afe961ff88e1b28074374eea19342ff010e9ec2350b883ebb6daf6fd84caa0bace08fff010280441fe8fc3e70797160405dea2daf308f104607fdb4ff008efa0d6d1d94d43bc0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108714f9e293a0ed9afd2250a3b525227e08f104607fcd74f008633cd801454f4fc50083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010c54d10927f27ba8e111d526dc6485e0108f104607fdb4ff0080a6635209395b6620083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010d281fc7c2bba2756ed4b8ed2a8e531e908f104607fdb4ff00897762a6d5293a14b0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by theStack** [comment]
Concept ACK

**Review by ryanofsky — COMMENTED** [review]
Updated 66bbbc09e9bb44546ac47829384cd3b088c5efd7 -> 615965cfd1ef1e0627d69970d99bdfedb9176833 ([`pr/initc.2`](https://github.com/ryanofsky/bitcoin/commits/pr/initc.2) -> [`pr/initc.3`](https://github.com/ryanofsky/bitcoin/commits/pr/initc.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/initc.2..pr/initc.3)) with suggested changes

**Review by KenG-pro — APPROVED** [review]

**Comment by MarcoFalke** [comment]
review ACK 615965cfd1ef1e0627d69970d99bdfedb9176833 🖱

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 615965cfd1ef1e0627d69970d99bdfedb9176833 🖱
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUj7CQv/Uj0PueOUes2yUiwCDP3lWhYevwJ0p5jaA85RtfTwsWnIAClUV0UmHAJh
cNpmWT73Dgg8Ku0jQbHO0cf9adIeKv8sbA3f+jQeStiz8FogOR6CWtgX4RzbLAWZ
W+x2Dq2ReHgToHggkQLWbEATGcZr2cQtub5NSjEwyfHnjIf8OvI7VOV7Ga8tsAWk
3oHSdjnWkfOcJLAkUZD9SF40fTjjPWFzI4KTwvp8SrQhL+uvuA4/3nk74idlCYE9
tYmWuiJ5MFocDOF5MVV/OtmWt+vfI7w2wbRYGu0N7LQU5lB9y2UgB8RzTmOCpKDz
qIrjsPdBNTiFcMOSYpJTE0s+Efuixi53nbHCWaJX2mgu7yzHZs/ktxw8NGeTUU2u
5nNs147u2YUaQ//aNcdbvRc7etvCT+do3N4zAYvNw9GNSHTjjPWT4uStX1qXMNnd
UqbUfMUGUTDjqgIcG8bpPX6WegkUIrCi7GZwbCLVf0FjXvDd4rWkxfS4Ijpy9rh/
/CfJrvyw
=XCNq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `460d9d89b06872cded3ec1cd1b35c0c9337114cc3b7071cebe084cad1768397b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108460d9d89b06872cded3ec1cd1b35c0c9337114cc3b7071cebe084cad1768397bf010f73b67ba6b9c524fe709b80dba2a635e08fff01047a40618677940dd682b882e1775518e08f10460811303f008807d9463aec63afa0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0108585f66a84c383622a24e4e4793fd81008f10460810528f008cee63d7da985f3400083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010bd3b798eceeffc895b094fd3ebe6de8608f10460811302f00865c18d2c0967c0220083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010c2db51b7061685b70cbfaea52596138e08f10460811302f00800f131a14c7b5dd50083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

File: src/init.cpp (hunk 6 above)
Comment (MarcoFalke) [reply src/init.cpp]: pubkey not addressed?

Comment (ryanofsky) [reply src/init.cpp]: re: https://github.com/bitcoin/bitcoin/pull/21732#discussion_r618106018

> pubkey not addressed?

Yes I dropped `<key.h>` but didn't see any `<pubkey.h>` so assumed it was mistake. Are we looking at the same file https://github.com/ryanofsky/bitcoin/blob/pr/initc.3/src/init.cpp?

Comment (MarcoFalke) [reply src/init.cpp]: Ah, so someone forgot to include it before the change. All good.

**Comment by practicalswift** [comment]
cr ACK 615965cfd1ef1e0627d69970d99bdfedb9176833: dimmed zebra looks correct

**Pull request merged**